dbusapi/tests/test_typeformatter.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# vim: ai ts=4 sts=4 et sw=4
+#
+# Copyright © 2017 Philip Withnall
+#
+# This library is free software; you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation; either version 2.1 of the License, or (at your option)
+# any later version.
+#
+# This library is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Unit tests for dbusapi.typeformatter
+"""
+
+
+from dbusapi.typeformatter import BasicTypeFormatter
+from dbusapi import types
+import unittest
+
+
+class TestBasicFormatterNormal(unittest.TestCase):
+    """Test normal formatting of types with the BasicTypeFormatter."""
+
+    # pylint: disable=invalid-name
+    def assertFormat(self, type_instance, formatted):  # noqa
+        formatter = BasicTypeFormatter()
+        self.assertEqual(formatter.format(type_instance), formatted)
+
+    def test_basic_types(self):
+        basic_type_map = {
+            types.Byte: 'Byte',
+            types.Boolean: 'Boolean',
+            types.Int16: 'Int16',
+            types.UInt16: 'UInt16',
+            types.Int32: 'Int32',
+            types.UInt32: 'UInt32',
+            types.Int64: 'Int64',
+            types.UInt64: 'UInt64',
+            types.Double: 'Double',
+            types.String: 'String',
+            types.ObjectPath: 'Object Path',
+            types.Signature: 'Signature',
+            types.Variant: 'Variant',
+            types.UnixFD: 'Unix FD',
+        }
+
+        for k, v in basic_type_map.items():
+            self.assertFormat(k(), v)
+
+    def test_array(self):
+        type_instance = types.Array()
+        type_instance.members.append(types.Byte())
+        self.assertFormat(type_instance, 'Array of [Byte]')
+
+    def test_struct(self):
+        type_instance = types.Struct()
+        type_instance.members.append(types.String())
+        type_instance.members.append(types.String())
+        type_instance.members.append(types.UInt16())
+        self.assertFormat(type_instance, 'Struct of (String, String, UInt16)')
+
+    def test_dict(self):
+        dict_instance = types.DictEntry()
+        dict_instance.members.append(types.String())
+        dict_instance.members.append(types.Variant())
+        array_instance = types.Array()
+        array_instance.members.append(dict_instance)
+
+        self.assertFormat(array_instance, 'Dict of {String: Variant}')
+
+    def test_invalid_dict(self):
+        """Test that a bare DictEntry (not in an Array) is not formatted"""
+        with self.assertRaises(AssertionError):
```

Comment:
I would put this right next to the line of code that is expected to assert, will spare you some head scratching when the DictEntry constructor will start to assert for whatever reason:)